CONSOLE-2508: Memsource automation

rebeccaalpert opened the pull request:
Added one script to handle project creation, file generation, and file upload for Memsource and a second to handle download and file import into OpenShift.

I used https://github.com/unofficial-memsource/memsource-cli-client instead of the Ansible collection because I couldn't get the collection to work. We can always circle back and switch to that later.

Documentation is in the files themselves; installing the memsource cli client and providing credentials is required for use. One script also requires the rename utility.

I also made some revisions to the other scripts in our little tool chain here.

Fixes https://issues.redhat.com/projects/CONSOLE/issues/CONSOLE-2508.

spadgett commented on frontend/i18n-scripts/memsource-upload.sh:
```diff
@@ -0,0 +1,41 @@
+
+echo "Creating jobs for exported PO files"
+
+memsource job create --filenames ${jaFiles[@]} --target-langs ja --project-id ${PROJECT_ID}
+memsource job create --filenames ${zhFiles[@]} --target-langs zh-cn --project-id ${PROJECT_ID}
+memsource job create --filenames ${koFiles[@]} --target-langs ko --project-id ${PROJECT_ID}
```
```suggestion
memsource job create --filenames "${jaFiles[@]}" --target-langs ja --project-id ${PROJECT_ID}
memsource job create --filenames "${zhFiles[@]}" --target-langs zh-cn --project-id ${PROJECT_ID}
memsource job create --filenames "${koFiles[@]}" --target-langs ko --project-id ${PROJECT_ID}
```

spadgett commented on frontend/i18n-scripts/memsource-upload.sh:
```diff
@@ -0,0 +1,41 @@
+do
+    case "${flag}" in
+        v) VERSION=${OPTARG};;
+        n) NAME=${OPTARG};;
```
```bash
    *) echo "usage: $0 [-v] [-n]" >&2
       exit 1;;
```

spadgett commented on frontend/i18n-scripts/memsource-upload.sh:
```diff
@@ -0,0 +1,41 @@
+
+echo "Exported all PO files"
+
+jaFiles=( po-files/ja/*.po )
```
We might consider adding a list of languages somewhere and making this a loop. That way adding or removing languages in the future is easier, and we're not duplicating commands.

rebeccaalpert commented:
Updated PR to add the download script. Let me know if you have feedback. I also addressed the feedback on script #1.

rebeccaalpert commented:
Just realized I didn't handle creating a new commit and PR in the script. I'll see if I can add that in.

rebeccaalpert commented:
Tried this per the gh CLI utility:
```git push -u origin HEAD
gh pr create --title "UI Translation update" --body "Adding latest translations for all languages from Memsource project id #$PROJECT_ID" --fill```

It seems like it's creating PRs with a ton of commits instead of just the desired one. Not sure what the issue is there.

rebeccaalpert commented:
I'm removing the automatic PR opening for now, but open to feedback.

spadgett commented on frontend/i18n-scripts/export-pos.sh:
```diff
@@ -2,22 +2,24 @@
-  yarn i18n-to-po -f $(basename "$f" .json) -l ko
+  for i in "${LANGUAGES[@]}"
+  do
+  yarn i18n-to-po -f $(basename "$f" .json) -l $i
```
It's a good practice to use quotes in case a filename could ever have special characters or spaces.

```suggestion
  yarn i18n-to-po -f "$(basename "$f" .json)" -l $i
```

spadgett commented on frontend/i18n-scripts/export-pos.sh:
```diff
@@ -2,22 +2,24 @@
-      yarn i18n-to-po -p "$d" -f $(basename "$f" .json) -l ko
+      for i in "${LANGUAGES[@]}"
+      do
+        yarn i18n-to-po -p "$d" -f $(basename "$f" .json) -l $i
```
```suggestion
        yarn i18n-to-po -p "$d" -f "$(basename "$f" .json)" -l "$i"
```

spadgett commented on frontend/i18n-scripts/memsource-download.sh:
```diff
@@ -0,0 +1,75 @@
+for i in "${LANGUAGES[@]}"
+do
+  COUNTER=0
+  CURRENT_PAGE=$(memsource job list --project-id $PROJECT_ID --target-lang $i -f value --page-number 0 -c uid | tr '\n' ' ')
```
I'd add quotes where you expand variables here and below.

```suggestion
  CURRENT_PAGE=$(memsource job list --project-id "$PROJECT_ID" --target-lang "$i" -f value --page-number 0 -c uid | tr '\n' ' ')
```

spadgett commented on frontend/i18n-scripts/memsource-download.sh:
```diff
@@ -0,0 +1,75 @@
+
+source ./i18n-scripts/languages.sh
+
+# Before running this tool, you must first install the unofficial Memsource CLI client: https://github.com/unofficial-memsource/memsource-cli-client#pip-install
+# You also have to configure it with your login info: https://github.com/unofficial-memsource/memsource-cli-client#configuration-red-hat-enterprise-linux-derivatives
+
+# On MacOS, you must also install rename, a GNU command line utility: brew install rename
```
I think this probably belongs in a README file under i18n-scripts

spadgett commented on frontend/i18n-scripts/memsource-download.sh:
```diff
@@ -0,0 +1,75 @@
+for i in "${LANGUAGES[@]}"
+do
+  cd $i
+  rename 's/%3D/=/' *
```
Can you give an example of what we get back and what we want?

spadgett commented on frontend/i18n-scripts/memsource-download.sh:
```diff
@@ -0,0 +1,75 @@
+done
+
+# Clean up PO file directory
+rm -rf downloaded_po_files
```
I would create a tmp dir for this instead with `mktemp`

spadgett commented on frontend/i18n-scripts/memsource-download.sh:
```diff
@@ -0,0 +1,75 @@
+echo Creating PR
+git add public/locales
+git add packages
+git commit -m "UI Translation update
```
I would use a semantic commit message like `chore(i18n): update translations`

spadgett replied:
We probably want to check that the git workspace is clean first to avoid accidentally committing other changes.

rebeccaalpert commented:
Addressed feedback and ran the download tool again. Seemed to work ok with the changes. Let me know if you have additional feedback.

dtaylor113 commented on frontend/i18n-scripts/README.md:
```diff
@@ -0,0 +1,56 @@
+You also have to [configure it with your Memsource login info](https://github.com/unofficial-memsource/memsource-cli-client#configuration-red-hat-enterprise-linux-derivatives).
+
+On MacOS, you must also install rename, a GNU command line utility, before running the download script.
+* You can use `brew install rename` to install rename if you need it
```
Does this work on Linux or Windows?  (do we care?)

rebeccaalpert replied:
It should work on Linux! The Memsource CLI Client should work on RHEL (see the docs) and I believe Linux has rename included.

rebeccaalpert replied:
I haven't tested it myself though.

dtaylor113 reviewed (approved):
/lgtm

dtaylor113 commented:
Nice job @rebeccaalpert!  Nice bit of scripting! :-)

spadgett reviewed (commented):
@rebeccaalpert A lot of variables are still missing quotes. Can you run `shellcheck` on the scripts and fix any errors? Thanks!

rebeccaalpert commented:
Fixed the stuff `shellcheck` called out.

spadgett commented:
Hey, @rebeccaalpert. There still seem to be warnings. Can you take another look? (Don't worry about the "not following" warnings.)

```
❯ shellcheck frontend/i18n-scripts/*.sh

In frontend/i18n-scripts/export-pos.sh line 5:
source ./i18n-scripts/languages.sh
       ^-------------------------^ SC1091: Not following: ./i18n-scripts/languages.sh was not specified as input (see shellcheck -x).


In frontend/i18n-scripts/export-pos.sh line 10:
  yarn i18n-to-po -f "$(basename "$f" .json)" -l $i
                                                 ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
  yarn i18n-to-po -f "$(basename "$f" .json)" -l "$i"


In frontend/i18n-scripts/export-pos.sh line 18:
    for f in $d/locales/en/* ; do
             ^-- SC2231: Quote expansions in this for loop glob to prevent wordsplitting, e.g. "$dir"/*.txt .


In frontend/i18n-scripts/export-pos.sh line 21:
        yarn i18n-to-po -p "$d" -f "$(basename "$f" .json)" -l $i
                                                               ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
        yarn i18n-to-po -p "$d" -f "$(basename "$f" .json)" -l "$i"


In frontend/i18n-scripts/languages.sh line 1:
LANGUAGES=( 'ja' 'zh-cn' 'ko')
^-- SC2148: Tips depend on target shell and yours is unknown. Add a shebang or a 'shell' directive.
^-------^ SC2034: LANGUAGES appears unused. Verify use (or export if used externally).


In frontend/i18n-scripts/memsource-download.sh line 5:
source ./i18n-scripts/languages.sh
       ^-------------------------^ SC1091: Not following: ./i18n-scripts/languages.sh was not specified as input (see shellcheck -x).


In frontend/i18n-scripts/memsource-upload.sh line 5:
source ./i18n-scripts/languages.sh
       ^-------------------------^ SC1091: Not following: ./i18n-scripts/languages.sh was not specified as input (see shellcheck -x).

For more information:
  https://www.shellcheck.net/wiki/SC2148 -- Tips depend on target shell and y...
  https://www.shellcheck.net/wiki/SC2034 -- LANGUAGES appears unused. Verify ...
  https://www.shellcheck.net/wiki/SC1091 -- Not following: ./i18n-scripts/lan...

```

rebeccaalpert commented:
@spadgett - I'm not getting those:
```
(.memsource) MacBook-Pro:frontend ralpert$ shellcheck -x i18n-scripts/memsource-download.sh
(.memsource) MacBook-Pro:frontend ralpert$ shellcheck -x i18n-scripts/memsource-upload.sh
```

spadgett commented on frontend/i18n-scripts/memsource-download.sh:
```diff
@@ -0,0 +1,74 @@
+  do
+    ((COUNTER++))
+    echo Downloading page "$COUNTER"
+    memsource job download --project-id "$PROJECT_ID" --output-dir "$DOWNLOAD_PATH"/"$i" --job-id "$CURRENT_PAGE"
```
nit: It's easier to quote the entire expression

```suggestion
    memsource job download --project-id "$PROJECT_ID" --output-dir "$DOWNLOAD_PATH/$i" --job-id "$CURRENT_PAGE"
```

spadgett commented on frontend/i18n-scripts/export-pos.sh:
```diff
@@ -2,22 +2,24 @@
 cd packages
 for d in */ ; do
-  if [ -d "$d/locales/en" ]
+  if [ -d "$d"/locales/en ]
```
I believe this was find as it was. You can quote the entire value, although what you have is OK, too.

spadgett commented on frontend/i18n-scripts/memsource-download.sh:
```diff
@@ -0,0 +1,74 @@
+  done
+done
+
+# Memsource gives us back file names like "console-app_package%3Dconsole-app."
+# We want names like "console-app_package=console-app" (with an equals sign)
+# in order to use our existing name-based file sorting mechanism.
```
Honestly, I think it would be better to just avoid the special characters in the filename altogether, particular if Memsource does weird things with them. Then we can also avoid needing the `rename` utility. Would a convention like `my-package__my-namespace` be simpler? (Similar to BEM names.)

rebeccaalpert replied:
I can make the change, but I'd have to rework two other scripts.

spadgett replied:
> I can make the change, but I'd have to rework two other scripts.

I think it's worth doing.

spadgett commented on frontend/i18n-scripts/memsource-upload.sh:
```diff
@@ -0,0 +1,35 @@
+
+echo "Creating project with title \"[OCP $VERSION] UI Localization - Sprint $SPRINT\""
+
+PROJECT_INFO=$(memsource project create --name "[OCP $VERSION] UI Localization - Sprint $SPRINT" --template-id 169304 -f json)
```
We should include the Git branch in here somewhere so that we can translate z-stream releases.

rebeccaalpert commented:
Addressed feedback. I adjusted naming scheme in a separate commit so we can drop it if needed. Note that this change will prevent us from using any of the scripts for dealing with translation files that we've already sent over under the original naming scheme (we have one batch currently in process).

spadgett commented on frontend/i18n-scripts/i18n-to-po.js:
```diff
@@ -84,9 +84,9 @@ function processFile(fileName, packageDir, language) {
-                `./../po-files/${language}/${path.basename(fileName)}_package=${
-                  packageDir.split('/')[0]
-                }.po`,
+                `./../po-files/${language}/${packageDir.split('/')[0]}_${path.basename(
```
I would make this two underscores `__` like BEM. Then we don't accidentally do the wrong thing if a package uses an underscore in its namespace. It makes it more obvious it's a meaningful separator, too.

spadgett commented on frontend/i18n-scripts/po-to-i18n.js:
```diff
@@ -14,12 +14,9 @@ function processFile(fileName, language) {
-    ? path.basename(fileName, '.po').split('_package=')[0]
-    : path.basename(fileName, '.po');
-  if (packageDir) {
+  const packageDir = path.basename(fileName, '.po').split('_')[0];
+  const newFileName = path.basename(fileName, '.po').split('_')[1];
```
```suggestion
  const [packageDir, newFileName] = path.basename(fileName, '.po').split('_');
```

spadgett reviewed (commented):
Thanks @rebeccaalpert

rebeccaalpert commented:
All set.

rebeccaalpert commented:
@spadgett; I squashed; let me know if this is good to go or if you have additional feedback. I used it to upload Sprint 200, but Sprint 199 will be in the old format and we'll have to revert the tools to the old naming scheme locally to import those.

spadgett reviewed (approved):
/lgtm

Thanks @rebeccaalpert.

spadgett commented:
/retest
